File: spring-data-requery/src/main/java/org/springframework/data/requery/repository/query/RequeryResultConverter.java
package org.springframework.data.requery.repository.query;

import io.requery.query.Tuple;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;

/**
 * RequeryResultConverter
 *
 * @author debop@coupang.com
 * @since 18. 6. 28
 */
@Slf4j
@UtilityClass
public class RequeryResultConverter {

    public static Object convertResult(Object result) {
        return convertResult(result, null);
    }

    public static Object convertResult(Object result, Object defaultValue) {

        log.trace("Convert result... result={}", result);
        try {
            if (result instanceof Tuple) {
                Tuple tuple = (Tuple) result;
                if (tuple.count() == 1)
                    return tuple.get(0);

                return tuple;
            }
            return result;

        } catch (Exception e) {
            log.warn("Fail to convert result[{}]. return [{}]", result, defaultValue);
            return defaultValue;
        }
    }
}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/repository/query/RequeryResultConverter.kt
package org.springframework.data.requery.repository.query

import io.requery.query.Tuple
import mu.KotlinLogging


object RequeryResultConverter {

    private val log = KotlinLogging.logger { }

    fun convert(result: Any?, defaultValue: Any? = null): Any? {

        log.trace { "Convert result. result=$result" }

        // TODO: 원하는 수형으로 변환하도록 해야 한다.
        return try {
            when(result) {
                is Tuple ->
                    when {
                        result.count() == 1 -> result.get(0)
                        else -> result
                    }
                else -> result
            }
        } catch(ignored: Exception) {
            log.warn(ignored) { "Fail to convert result[$result], return [$defaultValue]" }
            defaultValue
        }
    }
}